Add notebook on Hidden Linear Function Problem

fedimser opened the pull request:
This is one of problems, example for which was requested in #2620

fedimser commented:
build-docs failed with warning, treated as error:
/home/travis/build/quantumlib/Cirq/docs/notebooks/Hidden Linear Function Problem.ipynb:document isn't included in any toctree

Apparently I must add it to table of contents. Should I do it?

vtomole commented:
@fedimser Go ahead

vtomole commented:
@fedimser Oh. You put it in `docs` instead of `examples` Like https://github.com/quantumlib/Cirq/tree/master/examples/notebooks. I vote for leaving this here and moving the ones in `examples` to `docs`. Let's discus at the meeting today.

vtomole commented on docs/notebooks/Hidden Linear Function Problem.ipynb:
```diff
@@ -0,0 +1,522 @@
+   "source": [
+    "# Hidden Linear Function Problem\n",
+    "\n",
+    "In this notebook we consider a problem from the paper \"[Quantum advantage with shallow circuits](https://arxiv.org/pdf/1704.00690.pdf)\" and build a quantum circuit, which solves it, in Cirq. "
```
Don't link to the PDF. Link to the abstract: (https://arxiv.org/abs/1704.00690)

vtomole commented on docs/notebooks/Hidden Linear Function Problem.ipynb:
```diff
@@ -0,0 +1,522 @@
+    "\n",
+    "### The problem\n",
+    "\n",
+    "In this problem we consider a quadratic form of a binary vector and with binary coefficients (but additions and multiplications are evaluated modulo 4). Then we restrict this quadratic form, i.e. we allow to use only certain binary vectors as input. It turns out that under this restriction this quadratic form is equivalent to a linear function, i.e. it just evaluates dot product of input vec ...
```
Link quadratic form (https://en.wikipedia.org/wiki/Quadratic_form) and binary vector: https://en.wikipedia.org/wiki/Bit_array

vtomole commented on docs/notebooks/Hidden Linear Function Problem.ipynb:
```diff
@@ -0,0 +1,522 @@
+    "\n",
+    "Goal of this notebook is to introduce the reader to the problem and to show how Cirq can be used to solve it. We don't include proofs, but we refer the reader to corresponding lemmas in the original paper.\n",
+    "\n",
+    "### References\n",
```
Move these `References` to the bottom of the notebook.

fedimser commented:
@vtomole, I put it into cirq/docs/notebooks, because @dstrain115 said so in [this comment](https://github.com/quantumlib/Cirq/issues/2620#issuecomment-602199753).

fedimser commented:
Doug, can you please advise whether I should add link to this notebook to Table of Contents.
Currently in cirq/docs we don't have any case examples like this. Should I add a new section in cirq/docs/index.rst?
@dstrain115

dstrain115 commented:
> Doug, can you please advise whether I should add link to this notebook to Table of Contents.
> Currently in cirq/docs we don't have any case examples like this. Should I add a new section in cirq/docs/index.rst?

Can you make a new directory under docs/ called studies and move python notebook to docs/studies/hidden_linear_function.ipynb?

You should be able to add a section in the index.rst similar to this:

```
.. toctree::
    :maxdepth: 1
    :caption: Case Studies

   studies/hidden_linear_function.ipynb
```
Thanks.

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "#### 1. It's a problem without an oracle\n",
+    "\n",
+    "This problem is similar to a problem solved by [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm). It also finds coefficients of unknown linear function. But in the Bernstein-Vazirani algorithm this function is represented by an oracle. In this problem, the linear function is \"hidden\" in inputs $A$ and $b$.\n",
```
I think this paragraph needs some rewording since it is a bit awkward.  Here's my suggestion:

The [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm) also finds coefficients of an unknown linear function by representing by it as an oracle. By hiding the linear function in the inputs $A$ and $b$.\n", we avoid the use of an oracle.

You might also want to add why eliminating the oracle is useful.

vtomole replied:
> You might also want to add why eliminating the oracle is useful.

Yes. We really want to emphasize why this problem is important: It shows an unconditional separation between quantum and classical complexity classes. A result we usually don't get to have. In this case, a search problem in QNC^0 that is not in NC^0 (Link to https://en.wikipedia.org/wiki/NC_(complexity))

fedimser replied:
Rephrased.
Added the fact about complexity classes to next section about "Quantum circuits have advantage over classical when solving this problem"

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "In the first part, we solve this problem in a straightforward way, which follows directly from the problem definition. We will use this solution to verify our quantum solution in the next part. Also this part contains helper code to generate \"interesting\" instances of the problem.\n",
+    "\n",
+    "In the next part we solve this problem with Cirq. First, we write a code which builds a quantum circuit for solving arbitrary instances of the problem. Then we use Cirq's Clifford simulator to simulate this circuit. We do it for small instances and compare results to the bruteforce solution from the previous part. Then we solve the problem for a larger instance of the problem to demonstrate  ...
```
Should be a comma after "Then, ".  Brute force should be two words. Several instances of "bruteforce" elsewhere as well.

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "### Structure of this notebook\n",
+    "\n",
+    "In the first part, we solve this problem in a straightforward way, which follows directly from the problem definition. We will use this solution to verify our quantum solution in the next part. Also this part contains helper code to generate \"interesting\" instances of the problem.\n",
```
Comma after "Also,"

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+   "source": [
+    "## Preparation and bruteforce solution\n",
+    "\n",
+    "For small values of $n$ we can solve this problem with a trivial bruteforce solution. First, we need to build $\\mathcal{L}_q$ by checking for all $2^n$ binary vectors, which of them belongs to it (by definition). Then we need to try all possible $z \\in \\mathbb{F}_2^n$, and for each of them and for each $x \\in \\mathcal{L}_q$ check whether $q(x) = 2 z^T x$.\n",
```
"which of them belongs to it" is confusing.  I am not sure what you mean by this phrase.

fedimser replied:
Rephrased.

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+    "        \n",
+    "def find_interesting_problem(n, min_L_size):\n",
+    "    \"\"\"Generates \"interesting\" instance of the problem.\n",
+    "    \n",
```
Add explanation of the arguments

fedimser replied:
Done.

dstrain115 commented on docs/notebooks/Hidden_Linear_Function_Problem.ipynb:
```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "Note that: \n",
+    "\n",
+    "* We use Cirq S gate, whose matrix is $\\left(\\begin{smallmatrix}1 & 0\\\\0 & i\\end{smallmatrix}\\right)$. In the paper [1] matrix of S gate is defined as $\\left(\\begin{smallmatrix}1 & 0\\\\0 & -i\\end{smallmatrix}\\right)$. But for this problem it doesn't matter.\n",
```
Consider changing "Cirq S" to `cirq.S` to be more precise.

vtomole commented:
@fedimser  How about changing the filename to `HiddenLinearFunction` to be consistent with the notebooks already in `examples/notebooks`?

dstrain115 commented:
> I vote for leaving this here and moving the ones in `examples` to `docs`

Agreed.  I will move those over to docs in a later PR.  Thanks for pointing them out.

fedimser commented:
I addressed all comments. I put it in "Case studies" and added corresponding TOC entry, as Doug said in [this comment](https://github.com/quantumlib/Cirq/pull/2857#issuecomment-604122784).

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "\n",
+    "In this problem we consider a [quadratic form](https://en.wikipedia.org/wiki/Quadratic_form) of a [binary vector](https://en.wikipedia.org/wiki/Bit_array) and with binary coefficients (but additions and multiplications are evaluated modulo 4). Then we restrict this quadratic form, i.e. we allow to use only certain binary vectors as input. It turns out that under this restriction this quadrat ...
+    "\n",
+    "In other words, we have linear function, which is \"hidden\" inside a quadratic form.\n",
```
we have "a" linear function.

fedimser replied:
Done.

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "\n",
+    "Also define\n",
+    "\n",
+    "$$\\mathcal{L}_q = \\Big\\{x \\in  \\mathbb{F}_2^n : q(x \\oplus y) = (q(x) + q(y)) ~\\text{mod}~ 4 ~~ \\forall y \\in \\mathbb{F}_2^n \\Big\\}.$$\n",
```
I think we should add more explanation into this section.  What is the motivation behind these definitions?  Can we give the reader some intuition behind the math?

fedimser replied:
This problem is indeed contrived. I don't know how exactly authors of the paper came up with this particular problem statement. So I don't know what to write here. This problem doesn't represent any practcal problem from real life.

dstrain115 replied:
Can we give any intuition to the reader beyond a bunch of math that they need to read through?  I think it would be easier if we at least name the variables, e.g.  z is the hidden solution, etc.  That way, the reader can get a feel for what A, b, q, L_q, and z are.

fedimser replied:
I added verbal description for all variables in the end of this section.

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "In terms of complexity theory, this problem is in [QNC<sup>0</sup>](https://complexityzoo.uwaterloo.ca/Complexity_Zoo:Q#qnc0), but not in [NC<sup>0</sup>](https://en.wikipedia.org/wiki/NC_(complexity)), which shows that\n",
+    "QNC<sup>0</sup> $\\nsubseteq$ NC<sup>0</sup>.\n",
+    "\n",
+    "### Structure of this notebook\n",
```
This section seems out of place.  Should it be part of the introduction?

fedimser replied:
This was part of the introduction. I moved the problem statement in the separate chapter, so the introduction is shorter. Is it better now?

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "        n: dimension of the problem.\n",
+    "        min_L_size: minimal cardinality of subspace L.\n",
+    "        \n",
+    "    Returns instance of problem with gien `n`, such that size of \n",
```
gien => given.

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "        min_L_size: minimal cardinality of subspace L.\n",
+    "        \n",
+    "    Returns instance of problem with gien `n`, such that size of \n",
+    "    subspace `L_q` is at least `min_L_size`.\"\"\"\n",
```
I think this description should go before the "Args" section:

```
""Generates "interesting" instance of the problem.

    Returns instance of problem with gien `n`, such that size of 
    subspace `L_q` is at least `min_L_size`.

    Args:
        n: dimension of the problem.
        min_L_size: minimal cardinality of subspace L.
    """
```

fedimser replied:
Done.

dstrain115 commented on docs/studies/hidden_linear_function.ipynb:
```diff
@@ -0,0 +1,540 @@
+    "        problem.bruteforce_solve()\n",
+    "        if len(problem.L) >= min_L_size and not np.max(problem.A) == 0:\n",
+    "            return problem\n",
+    "    return None\n",
```
nit:  Should we raise an error instead of silently returning None?

fedimser replied:
This is auxillary code, so I don't think it matters.
And, strictly speaking, it's valid (not exceptional) situation if you didn't find "interesting" problem - for example, if min_L_size is to high.

dstrain115 commented:
Also, can you update the branch and see if it can be merged, now that the studies directory is in place?

fedimser commented:
PTAL and please merge if all is good.
@dstrain115

fedimser commented:
@vtomole, I addressed all your comments, can you please approve?

fedimser commented:
I just checked if I can merge it - and I can't.
Victory, can you please merge this PR?
@vtomole
